### ISSUE TITLE
grpc-tools build missing for arm64 macOS

### ISSUE THREAD

**Issue by bouk** [grpc/grpc#25755]
It's not possible to use grpc-tools on M1 macs right now, because there's no binary distribution of the Ruby gRPC compiler being shipped. Something will have to be done on the internal artifact build side to make this happen.

**Comment by jtattermusch** [comment]
we currently don't provide support for the arm64 M1 macs, since we have no hardware to test on. For now, you need to use rosetta emulation to run on the new arm64 macs.
We might add this support at some point, but there is no ETA.

**Comment by bouk** [comment]
@jtattermusch I made a PR so the x86_64 binary gets picked up at least: https://github.com/grpc/grpc/pull/25756

**Comment by sobrinho** [comment]
@jtattermusch what do you need to test it?

I'm sure the community can help, I have a M1 and I could help if you provide the instructions :)

**Comment by jtattermusch** [comment]
> @jtattermusch what do you need to test it?
> 
> I'm sure the community can help, I have a M1 and I could help if you provide the instructions :)

Thanks, but it's not about running a one-time test and be done with it. We need continous testing to be able to ensure things won't break in the future. So we need mac M1s integrated with our internal testing platform and that's far from trivial.

**Comment by jtattermusch** [comment]
related https://github.com/grpc/grpc/issues/26391

**Comment by sobrinho** [comment]
Maybe having instructions on how to compile from source would at least unblock to use grpc locally, until we have a definitive solution.

**Comment by nicolasnoble** [comment]
Also related: https://github.com/actions/virtual-environments/issues/2187

**Comment by jtattermusch** [comment]
> Maybe having instructions on how to compile from source would at least unblock to use grpc locally, until we have a definitive solution.

The way the grpc-tools gem is built by our CI is here:
https://github.com/grpc/grpc/blob/master/tools/run_tests/artifacts/build_package_ruby.sh

basically, you need to first build the protoc binaries and grpc_ruby_plugin for all supported platforms (this step is quite unrelated to the ruby build process, since you just need the binaries), then you put the binaries in the right location where gem build expects them and finally you simply run `gem build grpc-tools.gemspec`.
So besides building the protoc / protoc plugin binaries, building the grpc-tools gems is trivial.

Building grpc-tools for your own needs locally is absolutely doable (but obviously the gems built that way won't get any kind of official support from the grpc team).

**Comment by dweekly** [comment]
Hey folks, really would love to have first-class grpc-tools support for M1 macs since we're moving over from Intel-based to M1-based systems for doing our development and lean heavily on a grpc-tools stack for our workflow. I'm pretty sure I'm not the only one in this boat, so pretty please do find a way to support arm64!

**Comment by nicolasnoble** [comment]
@dweekly please see https://github.com/actions/virtual-environments/issues/2187

**Comment by dweekly** [comment]
@nicolasnoble Thanks so much; the linked thread indicates that GitHub is thinking that M1 support in Actions may be available sometime in 2022. https://github.com/actions/virtual-environments/issues/2187#issuecomment-939033228

I can appreciate the team's desire for quality in wanting to incorporate M1 Actions as part of your release flow, but it's also a non-super experience for the buyer of a modern, popular developer computer to not be able to use a core piece of software because one trillion dollar company is waiting for another trillion dollar company to do something.

Since I'm definitely not going to be the last person to run across this, perhaps it would be possible to have some sort of error pop up for those who encounter this with a link and guidance? Or to offer support in "preview"/beta/hand-cranked until full M1 GitHub actions are available for the team?

I'm appreciative of the hard work your team does *and* trying to be empathetic to devs wanting to use the grpc tools.

**Comment by sobrinho** [comment]
Looks like 1.42.0.pre1 was released, even include a version for arm64-darwin but it doesn't work yet

```
$ bin/rails db:migrate
rails aborted!
LoadError: dlopen(/Users/sobrinho/.gem/ruby/2.7.2/gems/grpc-1.42.0.pre1-universal-darwin/src/ruby/lib/grpc/2.7/grpc_c.bundle, 9): no suitable image found.  Did find:
	/Users/sobrinho/.gem/ruby/2.7.2/gems/grpc-1.42.0.pre1-universal-darwin/src/ruby/lib/grpc/2.7/grpc_c.bundle: mach-o, but wrong architecture
	/Users/sobrinho/.gem/ruby/2.7.2/gems/grpc-1.42.0.pre1-universal-darwin/src/ruby/lib/grpc/2.7/grpc_c.bundle: mach-o, but wrong architecture - /Users/sobrinho/.gem/ruby/2.7.2/gems/grpc-1.42.0.pre1-universal-darwin/src/ruby/lib/grpc/2.7/grpc_c.bundle
/Users/sobrinho/.gem/ruby/2.7.2/gems/activesupport-6.1.3.2/lib/active_support/dependencies.rb:332:in `require'
/Users/sobrinho/.gem/ruby/2.7.2/gems/activesupport-6.1.3.2/lib/active_support/dependencies.rb:332:in `block in require'
/Users/sobrinho/.gem/ruby/2.7.2/gems/activesupport-6.1.3.2/lib/active_support/dependencies.rb:299:in `load_dependency'
/Users/sobrinho/.gem/ruby/2.7.2/gems/activesupport-6.1.3.2/lib/active_support/dependencies.rb:332:in `require'
/Users/sobrinho/.gem/ruby/2.7.2/gems/grpc-1.42.0.pre1-universal-darwin/src/ruby/lib/grpc/grpc.rb:20:in `<top (required)>'
/Users/sobrinho/.gem/ruby/2.7.2/gems/grpc-1.42.0.pre1-universal-darwin/src/ruby/lib/grpc.rb:19:in `require_relative'
/Users/sobrinho/.gem/ruby/2.7.2/gems/grpc-1.42.0.pre1-universal-darwin/src/ruby/lib/grpc.rb:19:in `<top (required)>'
```

**Comment by sobrinho** [comment]
```
$ bundle update grpc
...
Installing grpc 1.42.0.pre1 (universal-darwin)
...

$ bin/rails db:migrate
rails aborted!
LoadError: dlopen(/Users/sobrinho/.gem/ruby/2.7.2/gems/grpc-1.42.0.pre1-universal-darwin/src/ruby/lib/grpc/2.7/grpc_c.bundle, 9): no suitable image found.  Did find:
	/Users/sobrinho/.gem/ruby/2.7.2/gems/grpc-1.42.0.pre1-universal-darwin/src/ruby/lib/grpc/2.7/grpc_c.bundle: mach-o, but wrong architecture
	/Users/sobrinho/.gem/ruby/2.7.2/gems/grpc-1.42.0.pre1-universal-darwin/src/ruby/lib/grpc/2.7/grpc_c.bundle: mach-o, but wrong architecture - /Users/sobrinho/.gem/ruby/2.7.2/gems/grpc-1.42.0.pre1-universal-darwin/src/ruby/lib/grpc/2.7/grpc_c.bundle
```

Although, if I do:

```
$ gem uninstall grpc
[uninstall all versions]

$ gem install grpc:1.42.0.pre1
Fetching grpc-1.42.0.pre1-arm64-darwin.gem
Successfully installed grpc-1.42.0.pre1-arm64-darwin
Parsing documentation for grpc-1.42.0.pre1-arm64-darwin
Installing ri documentation for grpc-1.42.0.pre1-arm64-darwin
Done installing documentation for grpc after 0 seconds
1 gem installed
```

It does install the correct architecture and then the app works.

**Comment by henriquebremenkanp** [comment]
@sobrinho I'm new to ruby and I'm not sure if it's the right solution, but maybe you need to add `arm64-darwin-20` to your Gemfile.lock platforms. I think it's what makes bundler to get the `arm64-darwin` instead of the `universal-darwin`. Not sure but maybe worth trying.

Btw, congratulations to the team on the newest [v1.42.0](https://github.com/grpc/grpc/releases/tag/v1.42.0) release!
I just noticed the latest gem is still released as [1.42.0.pre1](https://rubygems.org/gems/grpc/versions)

Does it only need a separate action to be triggered or is there anything left to be worked on?

**Comment by brodock** [comment]
to anyone who needs to get `grpc-tools` working on their development machines, you can temporarily set bundler platform to ruby which ignores any OS specific pre-built binary and rebuild directly from source:

bundle config set force_ruby_platform true

**Comment by kiskesis** [comment]
Any updates? Still see that issue
`node-pre-gyp ERR! stack Error: 404 status code downloading tarball https://node-precompiled-binaries.grpc.io/grpc-tools/v1.8.0/darwin-arm64.tar.gz
`

**Comment by brodock** [comment]
@kiskesis this is specific to ruby gem, I'm sure other language will also have the problem, but please find/open a different issue

**Comment by github-usr-1** [comment]
> basically, you need to first build the protoc binaries and grpc_ruby_plugin for all supported platforms

@jtattermusch Thanks for describing how to build grpc-tools.gemspec.  Can you also provide information on how to build the protoc binaries and grpc_ruby_plugin?

**Comment by sobrinho** [comment]
What about doing it manually when a new version is released?

It's tedious and error prone but better than not having it at all.

Maybe it can be compiled into a single script to build everything and on every release, we can do it ourselves and submit to a private repository or even provide the compiled version here for the owner to push to public rubygems.

**Comment by aspin** [comment]
In case this is helpful to anyone –– I do dev work from an M1 and the work around I ended up using was building everything from a Docker image. Something like this, built with the `--platform linux/amd64` flag include while building for compatibility:

```
FROM node:lts

# grpc_tools_node_protoc_ts for typescript bindings
RUN npm install -g grpc-tools grpc_tools_node_protoc_ts
RUN mkdir -p /home/node/out
RUN mkdir -p /home/node/in

WORKDIR "/home/node/in"
```

Then it's just a matter of mounting volumes in/out:

```
	docker run -v $(CURDIR)/server/proto:/home/node/out \
			   -v $(CURDIR)/proto:/home/node/in $(DOCKER_IMAGE) \
		grpc_tools_node_protoc \
			--plugin=protoc-gen-ts=/usr/local/bin/protoc-gen-ts \
			--ts_out=grpc_js:../out \
			--js_out=import_style=commonjs,binary:../out \
			--grpc_out=grpc_js:../out \
			api.proto
```

**Comment by lloeki** [comment]
> from an M1 [...] Something like this, built with the `--platform linux/amd64`

This only works if the Linux VM where docker is setup has `binfmt_misc` set up to run foreign `x86_64` binaries via either [Rosetta 2 for Linux](https://developer.apple.com/documentation/virtualization/running_intel_binaries_in_linux_vms_with_rosetta) (which requires using `Virtualization.framework`) or [QEMU](https://github.com/tonistiigi/binfmt) (which is slow).

**Comment by lloeki** [comment]
(reposting [my comment](https://github.com/grpc/grpc/issues/33954#issuecomment-1663968068) here for visibility)

I feel like it's more than that, the gem appears to break several rubygems expectations:

The gem is published as `ruby` (which means "pure ruby", or abusively "compatible with _any_ platform because it builds native components on the spot") but is in fact thin packaging for target-specific binaries:

```
$ env GEM_HOME=fake gem install grpc-tools
1 gem installed

$ ls -1 fake/gems/grpc-tools-1.56.2 fake/gems/grpc-tools-1.56.2/bin

fake/gems/grpc-tools-1.56.2:
README.md
bin
platform_check.rb
version.rb

fake/gems/grpc-tools-1.56.2/bin:
grpc_tools_ruby_protoc
grpc_tools_ruby_protoc_plugin
x86-linux
x86-windows
x86_64-linux
x86_64-macos
x86_64-windows
```

None but the one it ultimately runs on would be executable (without `binfmt_misc` trickery + a compatible userland), and so are dead weight.

Each one of these should be published as individual binary platform gems:

- `x86-linux`
- `x86_64-linux`
- `x86_64-darwin` (NOT `x86_64-darwin-22` or something, the missing platform version component acts as a wildcard)

For Windows I'm less certain, could be mswin or mingw thingies:

-https://github.com/rubygems/rubygems/blob/8af163866564c6a2afb9fd3cd51f452459dd9f06/lib/rubygems/platform.rb#L103-L127
- https://docs.ruby-lang.org/en/master/windows_md.html

And in addition, what this issue requests:

- `aarch64-linux`
- `arm64-darwin` (NOT `arm64-darwin-22` or something)

The `ruby` gem platform should be attempting to build binaries from source upon install, or worst case, be a shim.

In addition, this `case .. when` bit of code falls through to return `x86` on `aarch64-linux`:

```
  def PLATFORM.architecture
    host_cpu = RbConfig::CONFIG['host_cpu']

    # When we're on arm in macOS, we can rely on Rosetta and use the x86_64 binary
    return 'x86_64' if RbConfig::CONFIG['host_os'] =~ /darwin/ && host_cpu =~ /arm|aarch/

    case host_cpu
      when /x86_64/
        'x86_64'
      else
        'x86'
    end
  end
  ```

The `java` gem platform should include all binaries because it cannot be known which platform it will run on. But then I suppose JRuby is not a current target because the code above fails for JRuby.

For reference, here are some implementations:

- https://rubygems.org/gems/libv8-node/versions
- https://github.com/rubyjs/libv8-node/blob/4db6396edce91a5c17224c7c41a4b514dea1c85e/Rakefile
- https://github.com/DataDog/libddwaf-rb/blob/913c4083ab3db2f5b755079af171ec32ba06b015/Rakefile
- https://rubygems.org/gems/libddwaf/versions

**Comment by lloeki** [comment]
Some additional notes:

- [here's how to pick up the underlying platform triplet for JRuby](https://github.com/DataDog/libddwaf-rb/blob/913c4083ab3db2f5b755079af171ec32ba06b015/lib/datadog/appsec/waf.rb#L20-L56)
- if binaries do not run on musl†, I would recommend using `*-linux-gnu` to advertise it's only glibc compatible, otherwise `*-linux` would be picked by musl-built rubies as a fallback when there's no `*-linux-musl` gem available (which would get precedence)

† can be achieved by [statically linking C++ stdlib in and taking care that no glibc-isms are used](https://github.com/DataDog/libddwaf/tree/9caeac92a11ee7d43bf5d0d1453723409ac30f3f/docker/libddwaf#how-it-works)

**Comment by tebeco** [comment]
Yep that problem still persists with the upcoming M4

**Comment by tebeco** [comment]
> we currently don't provide support for the arm64 M1 macs, since we have no hardware to test on. For now, you need to use rosetta emulation to run on the new arm64 macs. We might add this support at some point, but there is no ETA.

somehow, @lloeki  / @jtattermusch someone has to review the idea of relying on Rosetta 2
Sonoma 14.5 broke stuff in that translation layer

TLDR: GRPC tool needs to be natively compliant with ARM processor, including Apple Silicon
this would cut the chain of dependency, hoping that translation layer don't break

**Comment by tebeco** [comment]
🦤
any news on google side for GRPC to be useable on ARM64 / Apple Silicon processor ?

Curious what's the google plan with mobile/desktop cpu moving to ARM64

also cheap cloud compute now runs on ARM64
it's honestly weird post M1 but we're now close to M4 and 2025. as a no ice in the field i fail to see any clear sign for support and things moving forward "actively"

**Comment by cannikin** [comment]
I was able to get the tools working inside a Docker container by telling it to enable i386 architecture support and then installing a required lib. This is a little standalone Ruby gRPC server that just runs a `server.rb` forever listening on port 4000 for requests from a separate Rails app I'm working on. Here's the complete Dockerfile:

```dockerfile
ARG RUBY_VERSION=3.3.4
FROM ruby:$RUBY_VERSION

WORKDIR /app
COPY . .

RUN dpkg --add-architecture i386
RUN apt update
RUN apt install -y libc6:i386
RUN bundle install

EXPOSE 4000

CMD ["ruby", "server.rb"]
```

I can also run the codebase in a devcontainer in VS Code (the devcontainer Dockerfile includes the same the three RUN commands that setup i386) and can verify that `grpc_tools_ruby_protoc -I . --ruby_out=. --grpc_out=. sample.proto` runs every time!
